lib/ansible/plugins/inventory/docker_swarm.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018, Stefan Heitmueller <stefan.heitmueller@gmx.com>
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: docker_swarm
+    plugin_type: inventory
+    authors:
+      - Stefan Heitmüller (stefan.heitmueller@gmx.com)
+    short_description: Ansible dynamic inventory plugin for Docker swarm nodes.
+    requirements:
+        - python >= 2.7
+        - Docker SDK for Python > 1.10.0
+    description:
+        - Reads inventories from the Docker swarm API.
+        - Uses a YAML configuration file docker_swarm.[yml|yaml].
+    options:
+        plugin:
+            description: the name of this plugin, it should always be set to 'docker_swarm' for this plugin to recognize it as it's own.
+            required: true
+            choices: ['docker_swarm']
+        host:
+            description: Socket of a Docker swarm manager node (tcp,unix).
+            required: true
+'''
+
+EXAMPLES = '''
+# docker_swarm.yml
+plugin: docker_swarm
+host: tcp://my-docker-host:2375
+'''
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.release import __version__
+
+try:
+    import docker
+    HAS_DOCKER = True
+except ImportError:
+    HAS_DOCKER = False
+
+
+class InventoryModule(BaseInventoryPlugin):
+    ''' Host inventory parser for ansible using Docker swarm as source. '''
+
+    NAME = "docker_swarm"
+
+    def _populate(self):
+        self.client = docker.DockerClient(base_url=self.get_option('host'))
+        self.inventory.add_group('all')
+        self.inventory.add_group('manager')
+        self.inventory.add_group('worker')
+        self.inventory.add_group('leader')
+        try:
+            self.nodes = self.client.nodes.list()
+            for self.node in self.nodes:
+                self.node_attrs = self.client.nodes.get(self.node.id).attrs
+                self.inventory.add_host(self.node_attrs['ID'])
+                self.inventory.add_host(self.node_attrs['ID'], group=self.node_attrs['Spec']['Role'])
+                self.inventory.set_variable(self.node_attrs['ID'], 'ansible_host', self.node_attrs['Status']['Addr'])
+                if 'ManagerStatus' in self.node_attrs:
+                    if self.node_attrs['ManagerStatus'].get('Leader'):
+                        self.inventory.add_host(self.node_attrs['ID'], group='leader')
+                if 'Labels' in self.node_attrs['Spec']:
```

Comment:
```suggestion
                if self.node_attrs['Spec'].get('Labels'):
```
This will prevent a crash when `'Labels' in self.node_attrs['Spec']`, but `self.node_attrs['Spec']['Labels']` is `None`.